plugins/incremental-ingestion-backend/README.md

```diff
@@ -0,0 +1,328 @@
+4. In cases when you can use `full` mutation, committing many entities into the processing pipeline fills up the processing queue and delays the processing of entities from other entity providers.
+
+We created the Incremental Entity Provider to address all of the above issues. The Incremental Entity Provider addresses these issues with a combination of `delta` mutations and a mark-and-sweep mechanism. Instead of doing a single `full` mutation, it performs a series of bursts. At the end of each burst, the Incremental Entity Provider performs the following three operations,
+
+1. Marks each received entity in the database.
+2. Annotates each entity with `backstage/incremental-entity-provider: <entity-provider-id>` annotation.
+3. Commits all of the entities with a `delta` mutation.
+
+Incremental Entity Providers will wait a configurable interval before proceeding to the next burst.
+
+Once the source has no more results, Incremental Entity Provider compares all entities annotated with `frontside/incremental-entity-provider: <entity-provider-id>` against all marked entities to determine which entities committed by same entity provider were not marked during the last ingestion cycle. All unmarked entities are deleted at the end of the cycle. The Incremental Entity Provider rests for a fixed internal before restarting the ingestion process.
+
+![Diagram of execution of an Incremental Entity Provider](https://user-images.githubusercontent.com/74687/185822734-ee6279c7-64fa-46b9-9aa8-d4092ab73858.png)
+
+This approach has the following benefits,
+
+1. Reduced ingestion latency - each burst commits entities which are processed before the entire list is processed.
+2. Stable pressure - each period between bursts provides an opportunity for the processing pipeline to settle without overwhelming the pipeline with a large number of unprocessed entities.
+3. Built-in retry / back-off - Failed bursts are automatically retried with a built-in back-off interval providing an opportunity for the data source to reset its rate limits before retrying the burst.
+4. Prevents orphan entities - Deleted entities are removed as with `full` mutation with a low memory footprint.
+
+## Requirements
+
+The Incremental Entity Provider backend is designed for data sources that provide paginated results. Each burst attempts to handle one or more pages of the query. The plugin will attempt to fetch as many pages as it can within a configurable burst length. At every iteration, it expects to receive the next cursor that will be used to query in the next iteration. Each iteration may happen on a different replica. This has several consequences:
+
+1. The cursor must be serializable to JSON (not an issue for most RESTful or GraphQL based APIs).
+2. The client must be stateless - a client is created from scratch for each iteration to allow distributing processing over multiple replicas.
+3. There must be sufficient storage in Postgres to handle the additional data.
```

Comment:
We might need to add here that it's only been tested with Postgres

Reply:
Done! The same considerations for storage would definitely apply to sqlite as well, but we never tested it with sqlite at all.